[TFLite] Update MultiplyByQuantizedMultiplier to use single-rounding instead of double-rounding

Tessil opened the pull request:
Hello,

This PR changes the rounding of the int32 `MultiplyByQuantizedMultiplier` methods to a single-rounding instead of a double-rounding in a way similar to what was done in https://github.com/google/ruy/pull/227 and adapts the tests to the change.

The new int32 `MultiplyByQuantizedMultiplier` adds a restriction on the quantized multiplier, it must now be a positive integer in the same way as in the int64 version of `MultiplyByQuantizedMultiplier`. This restriction also matches the one in the quantization scaling defined in the TOSA specification. To achieve that the SUB and DIV operators were slightly adapted to avoid any negative quantized multiplier.


Thibaut

Tessil committed:
Update MultiplyByQuantizedMultiplier to use single-rounding instead of double-rounding

Tessil committed:
Fix tests failing due to the rounding change in MultiplyByQuantizedMultiplier

Tessil committed:
Avoid the usage of negative quantized multipliers in the SUB operator

Tessil committed:
Avoid the usage of negative quantized multipliers in the DIV operator

Tessil committed:
Fix int16 SUB kernel to take into account the quantization scaling/zero-point

Tessil committed:
Update SUB operator tests to have different scaling in the inputs and outputs

The change provides a more in-depth testing of the SUB operator. Previously the same scaling was used for both inputs and for the output, a simple subtraction in the kernel without any rescaling would have made the tests pass.

Tessil committed:
Remove outdated comments in the SUB operator

renjie-liu commented on tensorflow/lite/kernels/sub_test.cc:
```diff
@@ -333,8 +333,8 @@ void QuantizedTestsActivationRELU_N1_TO_1() {
   for (int i = 0; i < inputs1.size(); ++i) {
-    QuantizedSubOpModel m({tensor_type, {1, 2, 2, 1}, -1.0, 1.0},
-                          {tensor_type, {1, 2, 2, 1}, -1.0, 1.0},
+    QuantizedSubOpModel m({tensor_type, {1, 2, 2, 1}, -0.9, 0.9},
```
why change this?

Tessil replied:
While changing the SUB kernel to avoid negative quantized multipliers an error slipped by as the quantized SUB kernel tests were not properly testing the operator. All the tests had the same scaling/zero-point for both inputs and the output which mean that a quantized kernel that just do a subtraction without taking into account the scaling would work. I fixed the int16 SUB kernel in 79c6ccc and adapted the tests in 2e94bea so that such errors can be catch.

renjie-liu commented on tensorflow/lite/kernels/internal/reference/sub.h:
```diff
@@ -376,19 +265,37 @@ inline void BroadcastSub16POTSlow(const ArithmeticParams& params,
-  TFLITE_DCHECK_LT(params.input1_offset, 256);
-  TFLITE_DCHECK_LT(params.input2_offset, 256);
+template <typename T, int N = 5>
+void BroadcastQuantSubSlow(const ArithmeticParams& params,
```
can we split this into other changes? thanks!

Tessil replied:
We could eventually separate the commits in the PR into multiple PRs as only the first commit fb9031aaac099feb3d58e5a59e2003890eef5982 changes the rounding. But the rounding change affects the tests and the DIV/SUB kernels. If we separate them in multiple PRs the CI will fail. I think it'd be easier to review each commit change one by one and still keep one PR.

renjie-liu replied:
I think we should do this incrementally: 1 pr to fix the sub/div tests scaling issue and another to update MultiplyByQuantizedMuliplier and adjust for rounding issue

renjie-liu commented on tensorflow/lite/kernels/internal/common.h:
```diff
@@ -181,46 +182,46 @@ inline int32_t MultiplyByQuantizedMultiplier(int64_t x,
+      right_shift_dup);
+
+  result.val[2] = vrshlq_s32(
+      vqdmulhq_s32(vshlq_s32(input_val.val[2], left_shift_dup), multiplier_dup),
```
why change this?

vqrdmulhq_s32 is rounding right?

Tessil replied:
Yes, vqrdmulhq_s32 is a multiplication with rounding which we replace with vqdmulhq_s32 which is a multiplication with truncation to avoid the double-rounding (with vrshlq_s32 as second rounding). 

This change is aligned to the change done in https://github.com/google/ruy/pull/227 (the discussion there provide more details) and matches the behaviour of the modified MultiplyByQuantizedMultiplier .

renjie-liu replied:
thanks for the explanation!

renjie-liu commented on tensorflow/lite/kernels/internal/optimized/neon_tensor_utils.cc:
```diff
@@ -143,53 +143,23 @@ inline int32_t AccumulateNeonLane(const int32x4_t lane) {
-  }
-#endif
+  result.val[0] = vrshlq_s32(
+      vqdmulhq_s32(vshlq_s32(input_val.val[0], left_shift_dup), multiplier_dup),
```
we will need the rounding here

Tessil replied:
See comment above.

Tessil committed:
Update optimized_ops::Quantize rescaling to single-rounding

renjie-liu commented on tensorflow/lite/kernels/internal/common.h:
```diff
@@ -138,31 +138,32 @@ inline void BiasAndClamp(float clamp_min, float clamp_max, int bias_size,
+inline int32_t MultiplyByQuantizedMultiplier(int32_t x,
+                                             int32_t quantized_multiplier,
+                                             int shift) {
+  assert(quantized_multiplier >= 0);
```
instead of assert, please use TFLITE_DCHECK whenever possible, thanks

Tessil replied:
Done, I also changed them in the int64 version as I initially used `assert` for coherence with this version.

renjie-liu reviewed (approved):
lgtm, please wait for others' opinions as well, thanks!

Tessil committed:
Revert SUB and DIV kernel changes

Tessil committed:
Replace assert by TFLITE_DCHECK

Tessil commented:
Thank you for the review. 

@renjie-liu I split the PR in two. The https://github.com/tensorflow/tensorflow/pull/50615 PR changes the SUB and DIV kernels to avoid the usage of negative quantized multiplier in `MultiplyByQuantizedMuliplier` and I left this one to concentrate on the `MultiplyByQuantizedMuliplier` rounding changes and the tests adaptation. The first one should be merged before this one for the tests to pass.

Tessil commented:
Looking into the failing tests (https://source.cloud.google.com/results/invocations/a9cc4ded-4646-4e73-9355-654127ead78c/targets) it seems there are still some rounding problems with the depth-wise convolution.

The DepthwiseConv kernel has a configurable rounding through the [DepthwiseConvOutputRounding](https://github.com/tensorflow/tensorflow/blob/aff4216fafd55a107ec617d534b5b49ef31f475a/tensorflow/lite/kernels/internal/reference/depthwiseconv_uint8.h#L47) enum. I was wondering if there was a reason for it as for what I know no other kernel has such configurable rounding? Should I just modify all the calls to use the same rounding or could I eventually remove this configurability altogether?

renjie-liu commented:
Sorry about the long hold, I think we can deprecate the `AwayFromZero` path but only take the `Upward` path.

The `Upward` path should be very similar to what you did, and we already allow the rounding difference for the test.

Tessil committed:
Adapt DepthwiseConv rounding to the MultiplyByQuantizedMultiplier rounding change

- DepthwiseConvRound<DepthwiseConvOutputRounding::kAwayFromZero> is replaced by the code used for the old MultiplyByQuantizedMultiplier
- DepthwiseConvRound<DepthwiseConvOutputRounding::kUpward> uses the new MultiplyByQuantizedMultiplier
- tflite::optimized_integer_ops::DepthwiseConvImpl, tflite::optimized_integer_ops::DepthwiseConvHybridImpl and tflite::reference_ops::DepthwiseConv uses DepthwiseConvOutputRounding::kUpward rounding instead of DepthwiseConvOutputRounding::kAwayFromZero for coherence between the rounding

jalexstark reviewed (commented):
I have looked at this quite a bit.

To help with review (this is an important and tricky change): Are we sure that in each instance 0.5 is added before truncation in order to have rounding rather than (overall) truncation?

In many cases we add the 0.5 to an accumulator at the start of arithmetic (since the accumulator is often seeded with a non-zero value anyway), so that the truncation at the end effectively performs rounding. Of course, this does not work if there is a non-shift multiplication at the end.

The reason for this is that I will try to trace through some of the computation, but I'd like to be sure that I don't have to repeat doing so. My mind is not currently "in this code" at present, so I am having to retrieve from old memories.

renjie-liu reviewed (changes requested):
let's wait for Alex's comments, thanks!

Tessil commented:
@renjie-liu I updated the PR and adapted the `DepthwiseConvRound` rounding to the change in `MultiplyByQuantizedMultiplier`. I also adapted some kernels that were using `DepthwiseConvOutputRounding::kAwayFromZero` to now use `DepthwiseConvOutputRounding::kUpward`. 

Note also that https://github.com/tensorflow/tensorflow/pull/50615#issuecomment-879788763 will need to be taken care of before testing and merging this PR.

@jalexstark Thank you for the review.

I don't know any kernel that initializes the accumulator with a quantized 0.5 value, do you eventually have an example? As you said it'll not work if there is a non-shift multiplication (i.e. `quantized_multiplier != 1`) and as the old `MultiplyByQuantizedMultiplier` was already doing a double-rounding it seems odd that some kernels would add an extra rounding by initializing the accumulator to a quantized 0.5.

jalexstark reviewed (commented):
Just one last thing from me.

Could someone who has this code available run tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/softmax_quantized_test.cc
with kTestsToRun much much larger (I suspect 10000 too slow, but do run for a few minutes)?

The reason behind this is that the numerics in the softmax prep routine were carefully worked out when they were written.

Tessil commented:
@jalexstark I tried with 10 000 and the tests go well but there are some failures with 100 000 though they also appear without the `MultiplyByQuantizedMultiplier` change.

jalexstark commented:
> @jalexstark I tried with 10 000 and the tests go well but there are some failures with 100 000 though they also appear without the `MultiplyByQuantizedMultiplier` change.

I suppose that we have to live with that. The difficulty has been that the burden of fuzz testing the numerics was a bit high, so the test size was reduced. Thanks.

Tessil committed:
Merge branch 'master' into toupstream/single_rounding

renjie-liu reviewed (commented):
looks this change will break micro tests as well, I think it's probably easier to duplicate the implementation then change the usage in separate prs, wdyt?

Tessil committed:
Merge branch 'master' into toupstream/single_rounding

Tessil committed:
Keep MultiplyByQuantizedMultiplier double-rounding as default rounding and put single-rounding behind a TFLITE_SINGLE_ROUNDING flag

Tessil committed:
Saturate the shift calculated by QuantizeMultiplier to 30 when TFL_SINGLE_ROUNDING is true

Tessil committed:
Saturate the PreprocessSoftmaxScaling multiplier to 1 << 30 instead of 1 << 31 to limit the size of the shift generated by QuantizeMultiplierGreaterThanOne

Tessil committed:
Test quantization_util_test with both single and double-rounding of MultiplyByQuantizedMultiplier

Tessil committed:
Add extra checks on the positivity of the quantized multiplier

Tessil committed:
Use (1 << 30) - 1 SoftmaxScaling saturation for single-rounding and (1 << 31) - 1 for double-rounding

jianlijianli commented:
Thanks Tessil. The PR looks great.

Tessil committed:
Keep MultiplyByQuantizedMultiplier tests with negative multipliers in double-rounding mode.

Tessil commented on tensorflow/lite/kernels/internal/quantization_util_test.cc:
```diff
@@ -433,17 +433,11 @@ TEST(QuantizationUtilTest, MultiplyByQuantizedMultiplierInt32) {
   EXPECT_EQ(quant_and_multiply(0, 0.1), 0);
   EXPECT_EQ(quant_and_multiply(1, 0), 0);
   EXPECT_EQ(quant_and_multiply(10000, 0.00097656), 10);
-  EXPECT_EQ(quant_and_multiply(10000, -0.00097656), -10);
```
Thanks, I added them back behind a `#if !TFLITE_SINGLE_ROUNDING`.

Tessil commented on tensorflow/lite/kernels/internal/reference/depthwiseconv_uint8.h:
```diff
@@ -68,6 +68,27 @@ inline int32_t DepthwiseConvRound(int32_t x, int32_t quantized_multiplier,
   return MultiplyByQuantizedMultiplier(x, quantized_multiplier, shift);
 }
 
+// Single-rounding MultiplyByQuantizedMultiplier
```
The single-rounding change in `MultiplyByQuantizedMultiplier` means that it now correspond to `DepthwiseConvOutputRounding::kUpward`. It'll be confusing later on if we don't add the change to the PR.

Tessil commented on tensorflow/lite/kernels/internal/quantization_util_test.cc:
```diff
@@ -498,12 +492,12 @@ TEST(QuantizationUtilTest, PreprocessSoftmaxScaling) {
 
   // If beta * scale is greater than fits in the number of integer bits, the
   // result is move near the maximum. Otherwise they quantize as expected.
-  // With 4 integer bits we can represent up to 16.0.
```
Not sure I understand the comment, nothing has changed for the double rounding case. All the tests were kept.

For the single-rounding, we now use 8.0 as scale to adapt to the saturation change in [PreprocessSoftmaxScaling](https://github.com/tensorflow/tensorflow/pull/50290/files#diff-f4849ed7b9b8d3d8daedafd91a17ae0b86e1690fb0417035d9ba71b5dcc8632fR294). The `beta * input_scale * (1 << (31 - input_integer_bits))` will here be `1.0 * 8.0 * (1 << (31 - 4)) = 1073741824.0` which is clamped to `(1LL << 30) - 1.0 = 1073741823.0`. With single-rounding it's `1.0 * 16.0 * (1 << (31 - 4)) = 2147483648.0`  clamped to `(1LL << 31) - 1.0 = 2147483647.0`. 

We can add an extra `EXPECT_THAT(quantize(1.0, 16.0, 4), Pair(2147483646, 30));` test for the single-rounding but not sure it has much interest as we already test the saturation with `8.0`.

Tessil commented:
Thanks @jianlijianli for the review, I added some comments.

gbaned commented:
@jianlijianli Can you please review this PR ? Thanks!

jianlijianli reviewed (approved):
Thanks Tessil.

gbaned commented:
@Tessil Can you please address Ubuntu Sanity errors? Thanks!

Tessil committed:
Rearrange cc_library arguments in BUILD file

Tessil commented:
@gbaned Thanks, the latest commit should fix it.

jianlijianli reviewed (approved):
Thanks Tessil.
